pep-0544.txt

```diff
@@ -0,0 +1,1006 @@
+this would be a regular (non-protocol) class that must implement ``Sized``.
+If ``Protocol`` is included in the base class list, all the other base classes
+must be protocols. A protocol can't extend a regular class.
+
+Alternatively, one can implement ``SizedAndCloseable`` like this, assuming
+the existence of ``SupportsClose`` from the example in `definition`_ section::
+
+  from typing import Sized
+
+  class SupportsClose(...): ...  # Like above
+
+  class SizedAndCloseable(Sized, SupportsClose, Protocol):
+      pass
+
+The two definitions of ``SizedAndClosable`` are equivalent.
+Subclass relationships between protocols are not meaningful when
+considering subtyping, since structural compatibility is
+the criterion, not the MRO.
+
+Note that rules around explicit subclassing are different from regular ABCs,
+where abstractness is simply defined by having at least one abstract method
+being unimplemented. Protocol classes must be marked *explicitly*.
+
+
+Generic and recursive protocols
+-------------------------------
+
+Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
+and ``Iterator`` are generic protocols. They are defined similar to normal
+non-protocol generic types::
+
+  T = TypeVar('T', covariant=True)
+
+  class Iterable(Protocol[T]):
+      @abstractmethod
+      def __iter__(self) -> Iterator[T]:
+          ...
+
+Note that ``Protocol[T, S, ...]`` is allowed as a shorthand for
+``Protocol, Generic[T, S, ...]``.
+
+Recursive protocols are also supported. Forward references to the protocol
+class names can be given as strings as specified by PEP 484. Recursive
+protocols will be useful for representing self-referential data structures
+like trees in an abstract fashion::
+
+  class Traversable(Protocol):
+      leaves: Iterable['Traversable']
+
+
+Using Protocols
+===============
+
+Subtyping relationships with other types
+----------------------------------------
+
+Protocols cannot be instantiated, so there are no values with
+protocol types. For variables and parameters with protocol types, subtyping
+relationships are subject to the following rules:
+
+* A protocol is never a subtype of a concrete type.
+* A concrete type or a protocol ``X`` is a subtype of another protocol ``P``
+  if and only if ``X`` implements all protocol members of ``P``. In other
+  words, subtyping with respect to a protocol is always structural.
+* Edge case: for recursive protocols, a class is considered a subtype of
+  the protocol in situations where such decision depends on itself.
+  Continuing the previous example::
+
+    class Tree(Generic[T]):
+        def __init__(self, value: T,
+                           leaves: 'List[Tree[T]]') -> None:
+            self.value = value
+            self.leafs = leafs
+
+    def walk(graph: Traversable) -> None:
+        ...
+    tree: Tree[float] = Tree(0, [])
+    walk(tree) # OK, 'Tree[float]' is a subtype of 'Traversable'
```

Comment:
Hm, I'm not sure I like this example. Isn't the problem more that `Traversible` ought to be generic but isn't in this example?